### ISSUE TITLE
Quick performance note

### ISSUE THREAD

**Issue by Yangqing** [ilkarman/DeepLearningFrameworks#2]
Hi - thanks for setting up the benchmark. I want to quickly put a performance note so folks are not surprised when seeing the perf numbers.

First, if one measure speed difference for basic ones like CNNs, something is wrong in how one use the frameworks :)

In this case, the major performance difference comes from I/O, not the framework itself. If you look at the TensorFlow and Caffe2 examples, the data is provided with a feed approach - this is usually bad for performance, and instead one should use a db or input iterator to do so. Under the hood, this makes prefetching and other optimizations possible and is particularly important for performance.

**Comment by botev** [comment]
Well, this a so-so. Depends on what you want to measure. If you want to measure overall performance than yes including the iterators/queues would be sensible to do. If you want to benchmark the performance of just the training function than making all of them get fed a dict is a right way to go. Overall, from my experience especially for big models like VGG to I/O is less than 2% of the performance, so I doubt to see that much difference with this.

**Comment by Yangqing** [comment]
It's VGG for Cifar, so it's actually a very small model, but w/e.

**Comment by ilkarman** [comment]
Yangqing thanks for taking the time comment! I definitely didn't intend to create a speed-comparison, especially ignoring disk/db pre-processing and image manipulation pipelines, running only on one GPU, with limited logging and using basic CNN layers. Instead I wanted a range of scripts that show the similarity between different frameworks and help people transition from one to another.

In this particular example the IO was negligible (for example MXNet's custom generator was only a few seconds faster), however what made a huge difference was one of the PRs:

`torch.backends.cudnn.benchmark=True # enables cudnn's auto-tuner`

This reduces the training-time from 351s to 165s

**Comment by ppwwyyxx** [comment]
Another thing that's unfair to TensorFlow is that your example uses NHWC layout for TensorFlow but NCHW for mxnet/pytorch/caffe2. 
See [TensorFlow performance guide](https://www.tensorflow.org/performance/performance_guide)

**Comment by ppwwyyxx** [comment]
Yet another thing that's unfair to TensorFlow:
As you said torch selects better cudnn algorithm after you enable the option. Similar to this, you should `export TF_ENABLE_WINOGRAD_NONFUSED=1` for TensorFlow. This also makes huge difference in performance.
It's not always enabled by default because they found bugs in certain versions of cudnn.

The above two changes brings the performance from 320s to 180s on a K40.

**Comment by ilkarman** [comment]
Thanks a lot for all of this help. With all of these changes we have converged to 78% for test-accuracy and around 170s for training-time with a small-ish variance. Hopefully the notebooks can act as examples of properly replicating the same CNN across all frameworks

**Comment by Yangqing** [comment]
Thanks for the note! Yes, I did a quick test and it seems that raw feeding - basically doing nothing but going through the network - creates a 2-4 second overhead for the overall training process. So indeed like you said it's not a dealbreaker.

We've experienced I/O being the main bottleneck for several of our in-production networks, so it would be nice to notify people that when one cares about top performances, using asynchronous I/O would help a lot :)
